src/precompiles/modexp.rs

```diff
@@ -60,96 +88,66 @@ impl Precompile for ModExp<Byzantium> {
         let cost = Self::required_gas(input)?;
         if cost > target_gas {
-            return Err(ExitError::OutOfGas);
+            Err(ExitError::OutOfGas)
+        } else {
+            let output = Self::run_inner(input)?;
+            Ok(PrecompileOutput::without_logs(cost, output))
         }
+    }
+}
 
+impl ModExp<Berlin> {
+    fn mul_complexity(base_len: u64, mod_len: u64) -> u64 {
+        let max_len = core::cmp::max(mod_len, base_len);
+        let words = max_len.div_ceil(&8);
+        words.saturating_mul(words)
+    }
+}
+
+impl Precompile for ModExp<Berlin> {
+    fn required_gas(input: &[u8]) -> Result<u64, ExitError> {
         let base_len = U256::from(&input[0..32]);
         let exp_len = U256::from(&input[32..64]);
         let mod_len = U256::from(&input[64..96]);
 
-        let base_len = base_len.as_usize();
-        let mut base_bytes = Vec::with_capacity(32);
-        for i in 0..base_len {
-            if 96 + i >= input.len() {
-                base_bytes.push(0u8);
-            } else {
-                base_bytes.push(input[96 + i]);
-            }
-        }
-
-        let exp_len = exp_len.as_usize();
-        let mut exp_bytes = Vec::with_capacity(32);
-        for i in 0..exp_len {
-            if 96 + base_len + i >= input.len() {
-                exp_bytes.push(0u8);
-            } else {
-                exp_bytes.push(input[96 + base_len + i]);
-            }
-        }
+        let mul = Self::mul_complexity(base_len.as_u64(), mod_len.as_u64());
+        let iter_count = Self::calc_iter_count(exp_len, base_len, &input).as_u64();
 
-        let mod_len = mod_len.as_usize();
-        let mut mod_bytes = Vec::with_capacity(32);
-        for i in 0..mod_len {
-            if 96 + base_len + exp_len + i >= input.len() {
-                mod_bytes.push(0u8);
-            } else {
-                mod_bytes.push(input[96 + base_len + exp_len + i]);
-            }
-        }
-
-        let base = BigUint::from_bytes_be(&base_bytes);
-        let exponent = BigUint::from_bytes_be(&exp_bytes);
-        let modulus = BigUint::from_bytes_be(&mod_bytes);
-
-        let output = {
-            let computed_result = base.modpow(&exponent, &modulus).to_bytes_be();
-            // The result must be the same length as the input modulus.
-            // To ensure this we pad on the left with zeros.
-            if mod_len > computed_result.len() {
-                let diff = mod_len - computed_result.len();
-                let mut padded_result = Vec::with_capacity(mod_len);
-                padded_result.extend(core::iter::repeat(0).take(diff));
-                padded_result.extend_from_slice(&computed_result);
-                padded_result
-            } else {
-                computed_result
-            }
-        };
-
-        Ok(PrecompileOutput::without_logs(cost, output))
+        Ok(core::cmp::max(200, mul * iter_count / 3))
```

Comment:
This arithmetic should be done in `U256` to prevent overflow (`mul` is bounded by `2^58` and `iter_count` is bounded by `2^40` so the whole expression is bounded by `2^98` which is too big for `u64`, but fine for `U256`). Then at the end we can round down to `u64`.